core/o11y/src/macros.rs

```diff
@@ -42,6 +42,28 @@ macro_rules! handler_trace_span {
     };
 }
 
+/// A macro to indicate non-fatal contract violation in the code.
+/// It is intended to be used in situations that should not happen
+/// such as some kind of invalid state or programmatic error, but not
+/// severe enough to crash production node with panic. Ideally we
+/// should aim at avoiding such situations in the first place with
+/// proper types, but in reality we have many implicit expectations 
+/// in our codebase and this macro should help highlighting those.
+/// The effects are as follows:
+/// * log an error to ensure the issue is reflected in node's logs
+/// * panic if in debug mode to uncover the underlying issue as soon
+///   as possible, achieved with `debug_assert!`
+#[macro_export]
+macro_rules! contract_violation {
+    (target: $target:expr, $($arg:tt)*) => {
+        tracing::error!(
+            target: $target,
+            $($arg)*
+        );
+        debug_assert!(false, $($arg)*);
```

Comment:
tracing::error takes complicated parameters that I don't think debug_assert supports? For example `tracing::error!(target: "abc", %display, ?debug, ...)`

So this macro kind of accepts a subset of `tracing::error!`. I wonder if we should just remove the target parameter, making it consistent with debug_assert!, panic!, println! etc. The target defaults to the current module anyway so I think that's OK (which is probably better than explicitly specifying a target).

Reply:
The best thing to do here would probably be to set up a temporary scoped tracing subscriber, format the event and the panic that out to the user – that way you get nicely readable asserts.

See [`tracing::subscriber::with_default`](https://docs.rs/tracing/latest/tracing/subscriber/fn.with_default.html) and [`tracing-subscriber::fmt::Layer::with_writer`](https://docs.rs/tracing-subscriber/latest/tracing_subscriber/fmt/struct.Layer.html#method.with_writer). From there you can have something like `impl io::Write for AssertDebugWriter` or something that explodes when it has a line written to it.